index.js

```diff
@@ -724,237 +495,227 @@ class XiaomiRoborockVacuum {
-        err
-      );
+      this.log.error(`ERR getFirmware | Failed getting the firmware version.`, err);
       throw err;
     }
   }
 
   get isCleaning() {
-    const status = this.device.property("state");
+    const status = this.device.property('state');
     return XiaomiRoborockVacuum.cleaningStatuses.includes(status);
   }
 
+
+  get isGoingToTarget() {
+    const status = this.device.property('state');
+    return 'unknown-16' === status;
+  }
+
+
+
+  //Changed by Mai
+  get isInPause() {
+    const status = this.device.property('state');
+    return 'paused' === status;
+  }
+
+
   async getCleaning() {
+    await this.ensureDevice('getCleaning');
+
     try {
-      const isCleaning = this.isCleaning;
-      this.log.info(
-        `INF getCleaning | ${this.model} | Cleaning is ${isCleaning}`
-      );
+      const isCleaning = this.isCleaning
+      this.log.info(`INF getCleaning | ${this.model} | Cleaning is ${isCleaning}`);
 
       return isCleaning;
     } catch (err) {
-      this.log.error(
-        `ERR getCleaning | Failed getting the cleaning status.`,
-        err
-      );
+      this.log.error(`ERR getCleaning | Failed getting the cleaning status.`, err);
       throw err;
     }
   }
 
+  //Changed by Mai
   async setCleaning(state) {
-    await this.ensureDevice("setCleaning");
-
+    await this.ensureDevice('setCleaning');
+    this.log.info(`ACT setCleaning | ${this.model} | Set cleaning to ${state}}`);
     try {
-      if (state && !this.isCleaning) {
-        // Start cleaning
-        this.log.info(
-          `ACT setCleaning | ${this.model} | Start cleaning, not charging.`
-        );
-        await this.device.activateCleaning();
-      } else if (!state) {
-        // Stop cleaning
-        this.log.info(
-          `ACT setCleaning | ${this.model} | Stop cleaning and go to charge.`
-        );
+      if (state && !this.isCleaning) { // Start cleaning
+        if (this.roomsToClean.length > 0) {
+          const refreshState = {
+            refresh: ['state'],
+            refreshDelay: 1000
+          };
+          await this.device.call('app_segment_clean', this.roomsToClean, refreshState);
+          this.log.info(`ACT setCleaning | ${this.model} | Start room cleaning for rooms ${this.roomsToClean}, the device is in state ${this.device.property('state')}`);
+
+        } else {
+          await this.device.activateCleaning();
+          this.log.info(`ACT setCleaning | ${this.model} | Start full cleaning, the device is in state ${this.device.property('state')}`);
+        }
+        return;
+      } else if (!state && (this.isCleaning || this.isInPause)) { // Stop cleaning
         await this.activateCharging(); // Charging works for 1st, not for 2nd
+        this.log.info(`ACT setCleaning | ${this.model} | Stop cleaning and go to charge, the device is in state ${this.device.property('state')}`);
+        //All rooms are removed from the list
+        this.roomsToClean.length = 0;
+        return;
```

Comment:
Actual change to include in a PR

I like this approach. I wonder though if we'd rather check the room switches over maintaining the `roomsToClean` array.